### DISCUSSION TITLE
Is there anything more that can be done to reduce the overhead of calling functions?

### DISCUSSION THREAD

**Discussion by jimbobmcgee** [PowerShell/PowerShell#15729] [Enhancements]
Considering the basest of contrived examples, below:

````
[CmdletBinding()]
Param(
    [Int64[]] $Iterations = (1000,10000,100000,1000000,10000000)
)
function script:func1 { # a standard function
    Param([Int64]$n)
    $script:value += $n
}

$script:func2 = { # a script block, which I can .Invoke() or .InvokeReturnAsIs()
    Param([Int64]$n)
    $script:value += $n
}

$script:func3 = [Action[Int64]] { # a script block coerced to an Action<>
    Param([Int64]$n)
    $script:value += $n
}

$cs = (Add-Type -Language CSharpVersion3 -PassThru -TypeDefinition @"
    public static class Lib_$([Guid]::NewGuid().ToString("N"))
    {   // a c# function, which I can passthru and call directly
        public static void DoIt(ref long orig, long n) 
        {
            orig += n;
        }
    }
"@)[0]

foreach ($script:iter in $Iterations) {
    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'Unrolled'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                $script:value += $i     # just don't call a function
            }
        })
    })

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'func1'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                script:func1($i)
            }
        })
    })

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'func2'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                $script:func2.Invoke($i)
            }
        })
    })

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'func2-asis'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                $script:func2.InvokeReturnAsIs($i)
            }
        })
    })

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'func3'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                $script:func3.Invoke($i)
            }
        })
    })

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'csfunc'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                $cs::DoIt([ref]$script:value, $i)
            }
        })
    })
}
````

...the results of this may be surprising:

````
Method     Timing           Iterations
------     ------           ----------
Unrolled   00:00:00.0093216       1000
func1      00:00:00.0463300       1000
func2      00:00:00.0260111       1000
func2-asis 00:00:00.0146741       1000
func3      00:00:00.0230474       1000
csfunc     00:00:00.0134276       1000
Unrolled   00:00:00.0168714      10000
func1      00:00:00.3295666      10000
func2      00:00:00.0773167      10000
func2-asis 00:00:00.0740876      10000
func3      00:00:00.0714550      10000
csfunc     00:00:00.0291256      10000
Unrolled   00:00:00.1600401     100000
func1      00:00:03.1881152     100000
func2      00:00:00.7301056     100000
func2-asis 00:00:00.7108403     100000
func3      00:00:00.7071193     100000
csfunc     00:00:00.2571653     100000
Unrolled   00:00:01.5404970    1000000
func1      00:00:31.6345677    1000000
func2      00:00:07.4991554    1000000
func2-asis 00:00:07.4149348    1000000
func3      00:00:07.2059518    1000000
csfunc     00:00:02.6307728    1000000
Unrolled   00:00:15.8265850   10000000
func1      00:05:19.1938283   10000000
func2      00:01:14.3032956   10000000
func2-asis 00:01:10.7433413   10000000
func3      00:01:10.2432960   10000000
csfunc     00:00:25.4396110   10000000
````

![image](https://user-images.githubusercontent.com/12205864/124676611-c5dde580-deb6-11eb-8267-76e3a7c46376.png)

My guess is that the woeful performance of `func1` is probably due to parameter binding but---at least to me---the overhead of calling a "normal" function is disappointing, especially after seeing a concerted effort to make a larger script more readable plummett performance by as much as *20x* -- significantly more if there were multiple function calls per iteration.

Appreciating that rewriting the entire script in C# is...*less than ideal*...are there any (preferably backward-compatible) approaches, in general, that I have missed which might improve upon the `func2` / `func3` best-case pure-Powershell approaches?

I'm personally stuck with running in Win PS -- readings were taken from Win10/PS5.1 -- I can't shift my current stack away from WinPS at the moment but, for knowledge's sake, is there anything in PS6/7+ that significantly improves on this?

Or is splitting work into readable functions just a write-off?

**Comment by iSazonov** [comment]
You are more likely to encounter worse performance on PS7. No improvements have been made in parameter bindings.

**Comment by SeeminglyScience** [comment]
The fastest advisable way will probably be `& $scriptBlock` by a lot I'd guess.

That said, something with `SteppablePipeline` will probably be the fastest overall assuming you reuse the same pipe. Not exactly a tested use case so don't be surprised if you run into strange issues though.


```powershell
using namespace System.Management.Automation

$script:value = 0L

$script:pipe = {
    & {
        [CmdletBinding()]
        param(
            [Parameter(ValueFromPipeline)]
            [int64] $i
        )
        process {
            $script:value += $i
        }
    }
}.GetSteppablePipeline([CommandOrigin]::Internal)

$script:pipe.Begin(<# expectingInput: #> $true)

for ($i = 0L; $i -lt 1000L; $i++) {
    $script:pipe.Process($i)
}
```

Also FWIW there's a huge amount of pit falls when benchmarking (especially very small benchmarks like these) in PowerShell.  Dot sourcing, the scope you pull access variables from, and a whole lot of other factors can make something look like worse performance than it actually is in a real world script.

**Reply by jimbobmcgee** [reply]
> Also FWIW there's a huge amount of pit falls when benchmarking...

Appreciated, but the effect was happening in my much larger scripted process (looping through a lot of DB rows, performing much more slowly after the initial loop innards were split into readable functions), and I wanted to post a minimal example to recreate.

> The fastest advisable way will probably be & $scriptBlock by a lot I'd guess.

I initially discounted `&` (and `Invoke-Command`, for that matter), thinking any gains I got would be lost in the extra call.  I've added `&` into the rudimentary benchmark and---unless I'm doing it wrong---see timings between the function call (func1) and the `$script.Invoke()` (func2).

> That said, something with SteppablePipeline will probably be the fastest overall assuming you reuse the same pipe

I'd not heard of this (and the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.scriptblock.getsteppablepipeline?view=powershellsdk-7.0.0) leave something to be desired).  I see similar (albeit slightly higher) timings to `$script.Invoke()`.

Additional tests (based on @SeeminglyScience's comments):
````
    Write-Host "&func2: $script:iter"
    Write-Output (New-Object PSObject -Property @{
        'Method'        = '&func2'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                &$script:func2 $i
            }
        })
        'Result'        = $script:value
    })

$script:pipe = {
    & {
        [CmdletBinding()]
        Param([Parameter(ValueFromPipeline=$true)][Int64]$n)
        Process {
            $script:value += $n
        }
    }
}.GetSteppablePipeline([Management.Automation.CommandOrigin]::Internal)

    Write-Host "pipe: $script:iter"
    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'pipe'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            $script:pipe.Begin($true)
            foreach ($i in (1..$script:iter)) {
                $script:pipe.Process($i)
            }
        })
        'Result'        = $script:value
    })
````

Current results (I added the Result column as a checksum, but it is outside of the timing):
````
Method             Result Timing           Iterations
------             ------ ------           ----------
Unrolled           500500 00:00:00.0078820       1000
func1              500500 00:00:00.0411375       1000
func2              500500 00:00:00.0229343       1000
func2-asis         500500 00:00:00.0235554       1000
&func2             500500 00:00:00.0299249       1000
func3              500500 00:00:00.0213464       1000
pipe               500500 00:00:00.0149422       1000
csfunc             500500 00:00:00.0113829       1000
Unrolled         50005000 00:00:00.0185185      10000
func1            50005000 00:00:00.3354716      10000
func2            50005000 00:00:00.0820472      10000
func2-asis       50005000 00:00:00.0789145      10000
&func2           50005000 00:00:00.2242624      10000
func3            50005000 00:00:00.0784974      10000
pipe             50005000 00:00:00.0910078      10000
csfunc           50005000 00:00:00.0305709      10000
Unrolled       5000050000 00:00:00.1738001     100000
func1          5000050000 00:00:03.2735234     100000
func2          5000050000 00:00:00.8045826     100000
func2-asis     5000050000 00:00:00.7438697     100000
&func2         5000050000 00:00:02.2146113     100000
func3          5000050000 00:00:00.7475501     100000
pipe           5000050000 00:00:00.8722211     100000
csfunc         5000050000 00:00:00.2763037     100000
Unrolled     500000500000 00:00:01.6654439    1000000
func1        500000500000 00:00:32.8118375    1000000
func2        500000500000 00:00:08.3787069    1000000
func2-asis   500000500000 00:00:07.9270452    1000000
&func2       500000500000 00:00:21.9441235    1000000
func3        500000500000 00:00:07.5450989    1000000
pipe         500000500000 00:00:08.9639209    1000000
csfunc       500000500000 00:00:02.9834846    1000000
Unrolled   50000005000000 00:00:17.0318852   10000000
func1      50000005000000 00:05:29.0132837   10000000
func2      50000005000000 00:01:18.8782519   10000000
func2-asis 50000005000000 00:01:15.2773396   10000000
&func2     50000005000000 00:03:41.8039564   10000000
func3      50000005000000 00:01:15.0565614   10000000
pipe       50000005000000 00:01:27.7718384   10000000
csfunc     50000005000000 00:00:27.4932446   10000000
````

![image](https://user-images.githubusercontent.com/12205864/124780338-6cb79580-df3a-11eb-9d0f-c0fddde50fc5.png)

**Comment by jimbobmcgee** [comment]
Has there been any suggestion and/or advancement in a MSIL JIT/compile routine for Powershell code?  I saw PS2EXE, but that looks like it bundles a host and feeds your script to it, rather than *compiling*, per se...

**Reply by SeeminglyScience** [reply]
It already is compiled after a certain number (~30 or so) of invocations. AOT compilation wouldn't buy much other than some minor start up costs.

**Reply by iSazonov** [reply]
Don't forget about PowerShell classes.

**Reply by jimbobmcgee** [reply]
@SeeminglyScience - the pure PS function is compiled after ~30 invocations?  Does that improve the parameter binder time *calling* the function (e.g. inside a tight loop) or "just" the time taken once *inside* the function?  If I had a short-circuiting conditional return at the top of my function, could I spin it ~30 times at the start to trigger the compile, then enjoy a high-speed invocation later, or would it only compile the short-circuited path previously seen?

@iSazonov  -  I'll be honest, I know nothing about PowerShell classes, but I bet, as soon as I start trying to use them, I'll end up on on a server running a PS that can't understand them...

**Reply by iSazonov** [reply]
If you use Windows PowerShell the latest supported version is 5.1 and if you create a PS class it will work on every system where PS 5.1 is installed.

**Reply by jimbobmcgee** [reply]
@iSazonov - the latest *supported* version might be 5.1, but the latest *deployed* version is seldom thus.  I've seen entirely "up to date" Windows systems, in the wild, with 2.0, 3.0, etc.  And, with the fragmentation of .NET, I don't see that changing, any time soon.

...but that's a separate discussion  :-)  I will look into classes in 5.1, and see if I they improve anything with regards to function call overhead.

**Reply by SeeminglyScience** [reply]
> @SeeminglyScience - the pure PS function is compiled after ~30 invocations?

Any scriptblock, yeah.

> Does that improve the parameter binder time _calling_ the function (e.g. inside a tight loop) or "just" the time taken once _inside_ the function?

The latter.

> If I had a short-circuiting conditional return at the top of my function, could I spin it ~30 times at the start to trigger the compile, then enjoy a high-speed invocation later, or would it only compile the short-circuited path previously seen?

Technically yeah, but if 30 invocations is a meaningful portion of total run time then you are not likely at a scale where interpreted vs compiled makes any noticeable difference.

**Reply by jimbobmcgee** [reply]
@SeeminglyScience 

> Technically yeah, but if 30 invocations is a meaningful portion of total run time then you are not likely at a scale where interpreted vs compiled makes any noticeable difference.

Agreed, and---since my pain is in the function call, not the implementation (although I'm sure the implementation leaves much to be desired)---not going to improve my runtime.  Was just trying to get a feel for whether it was sensitive to branching.

**Comment by jimbobmcgee** [comment]
So, it seems the route that I've gone (`{...}.Invoke()`) is probably the best I'm going to get for now, without a drastic refactor to C#, or a swathe of copy-paste.  Having done so, I note some idiosyncrasies, such as a reduced stack trace in errors.

Happy to open this up to ways in which further development of PowerShell could improve on the situation, especially in script/module code...

**Reply by AltitudeApps** [reply]
I think PowerShell can already deliver the performance you are looking for, see below.  (And the overhead for parameter binding is not significant, I don't think)

**Reply by AltitudeApps** [reply]
Actually, I take that back, the overhead for parameter binding can indeed be significant.

**Comment by AltitudeApps** [comment]
Try using the pipeline, and omit an explicit parameter block:

```pwsh
[CmdletBinding()]
Param(
    [Int64[]] $Iterations = (1000,10000,100000,1000000) #,10000000)
)

$value = 999

function func0 {
  # this is the way
  process {
    $value += $_
  }
}

function func0B {
  param(
    [parameter( mandatory=$true,ValueFromPipeline=$true)]
    $x
  )
  process {
    $value += $x
  }
}

function script:func1 { # a standard function
    Param([Int64]$n)
    $script:value += $n
}

$script:func2 = { # a script block, which I can .Invoke() or .InvokeReturnAsIs()
    Param([Int64]$n)
    $script:value += $n
}

$script:func3 = [Action[Int64]] { # a script block coerced to an Action<>
    Param([Int64]$n)
    $script:value += $n
}

$cs = (Add-Type -Language CSharp -PassThru -TypeDefinition @"
    public static class Lib_$([Guid]::NewGuid().ToString("N"))
    {   // a c# function, which I can passthru and call directly
        public static void DoIt(ref long orig, long n) 
        {
            orig += n;
        }
    }
"@)[0]

$sbCsFuncPipe = `
    {
      process {
          $cs::DoIt([ref]$script:value, $_ ) 
      }
    }

foreach ($script:iter in $Iterations) {

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'Unrolled  - make this wide ...'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                $script:value += $i     # just don't call a function
            }
        })
    })

    Write-Output "value: $($script:value)"

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'func zero'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            $func = $function:func0

            $thisMany = 1 .. $script:iter
            # use the pipeline
            $thisMany | . $func
            # 1 .. $script.iter  | . $func

        })
    })

    Write-Output "value: $($script:value)"

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'func zero B'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            $func = $function:func0B

            # use the pipeline, with function that has a named parameter
            $thisMany = 1 .. $script:iter
            $thisMany | . $func
            # ( 1 .. $script.iter ) | . $func

        })
    })

    Write-Output "value: $($script:value)"

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'func zero B via x'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            $func = $function:func0B


            # use the pipeline, with function that has a named parameter,
            # but via `ForEach-Object`, as if you didn't know that the
            # function accepted pipeline input
            $thisMany = 1 .. $script:iter
            $thisMany | Foreach-Object { . $func -x $_ }

            # ( 1 .. $script.iter ) | Foreach-Object { . $func -x $_ }

        })
    })

    Write-Output "value: $($script:value)"

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'csfunc'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L
            foreach ($i in (1..$script:iter)) {
                $cs::DoIt([ref]$script:value, $i)
            }
        })
    })

    Write-Output "value: $($script:value)"

    Write-Output (New-Object PSObject -Property @{
        'Method'        = 'csfunc via pipeline'
        'Iterations'    = $script:iter
        'Timing'        = (Measure-Command {
            [Int64] $script:value = 0L

            # use the csharp function indirectly, via a powershell function
            # that accepts pipeline input
            $thisMany = 1 .. $script:iter
            $thisMany | & $sbCsFuncPipe

        })
    })

    Write-Output "value: $($script:value)"

    Write-Output "-----------`n"

    # Don't even bother with these:

    # Write-Output (New-Object PSObject -Property @{
    #     'Method'        = 'func zero B via foreach'
    #     'Iterations'    = $script:iter
    #     'Timing'        = (Measure-Command {
    #         [Int64] $script:value = 0L
    #         $func = $function:func0B

    #         foreach ($i in (1..$script:iter)) {
    #             & $func $i
    #         }
    #     })
    # })

    # Write-Output (New-Object PSObject -Property @{
    #     'Method'        = 'func1C'
    #     'Iterations'    = $script:iter
    #     'Timing'        = (Measure-Command {
    #         [Int64] $script:value = 0L
    #         $func = $function:func1
    #         foreach ($i in (1..$script:iter)) {
    #             & $func $i
    #         }
    #     })
    # })



    # Write-Output (New-Object PSObject -Property @{
    #     'Method'        = 'func1'
    #     'Iterations'    = $script:iter
    #     'Timing'        = (Measure-Command {
    #         [Int64] $script:value = 0L
    #         foreach ($i in (1..$script:iter)) {
    #             script:func1($i)
    #         }
    #     })
    # })

    # Write-Output (New-Object PSObject -Property @{
    #     'Method'        = 'func1B'
    #     'Iterations'    = $script:iter
    #     'Timing'        = (Measure-Command {
    #         [Int64] $script:value = 0L
    #         foreach ($i in (1..$script:iter)) {
    #             script:func1 $i
    #         }
    #     })
    # })



    # Write-Output (New-Object PSObject -Property @{
    #     'Method'        = 'func2'
    #     'Iterations'    = $script:iter
    #     'Timing'        = (Measure-Command {
    #         [Int64] $script:value = 0L
    #         foreach ($i in (1..$script:iter)) {
    #             $script:func2.Invoke($i)
    #         }
    #     })
    # })

    # Write-Output (New-Object PSObject -Property @{
    #     'Method'        = 'func2-asis'
    #     'Iterations'    = $script:iter
    #     'Timing'        = (Measure-Command {
    #         [Int64] $script:value = 0L
    #         foreach ($i in (1..$script:iter)) {
    #             $script:func2.InvokeReturnAsIs($i)
    #         }
    #     })
    # })

    # Write-Output (New-Object PSObject -Property @{
    #     'Method'        = 'func3'
    #     'Iterations'    = $script:iter
    #     'Timing'        = (Measure-Command {
    #         [Int64] $script:value = 0L
    #         foreach ($i in (1..$script:iter)) {
    #             $script:func3.Invoke($i)
    #         }
    #     })
    # })

}
```

**Reply by AltitudeApps** [reply]
```txt
Method                         Timing           Iterations
------                         ------           ----------
Unrolled  - make this wide ... 00:00:00.0146774       1000
value: 500500
func zero                      00:00:00.0348789       1000
value: 500500
func zero B                    00:00:00.0381645       1000
value: 500500
func zero B via x              00:00:00.1808342       1000
value: 500500
csfunc                         00:00:00.0214909       1000
value: 500500
csfunc via pipeline            00:00:00.0387765       1000
value: 500500
-----------

Unrolled  - make this wide ... 00:00:00.0404534      10000
value: 50005000
func zero                      00:00:00.0458774      10000
value: 50005000
func zero B                    00:00:00.2083047      10000
value: 50005000
func zero B via x              00:00:01.9403793      10000
value: 50005000
csfunc                         00:00:00.0772508      10000
value: 50005000
csfunc via pipeline            00:00:00.0888102      10000
value: 50005000
-----------

Unrolled  - make this wide ... 00:00:00.3898073     100000
value: 5000050000
func zero                      00:00:00.4523467     100000
value: 5000050000
func zero B                    00:00:02.0723845     100000
value: 5000050000
func zero B via x              00:00:17.0378065     100000
value: 5000050000
csfunc                         00:00:00.7726129     100000
value: 5000050000
csfunc via pipeline            00:00:00.8940438     100000
value: 5000050000
-----------

Unrolled  - make this wide ... 00:00:03.8376150    1000000
value: 500000500000
func zero                      00:00:04.4173799    1000000
value: 500000500000
func zero B                    00:00:20.3821910    1000000
value: 500000500000
func zero B via x              00:03:01.2116492    1000000
value: 500000500000
csfunc                         00:00:08.7834210    1000000
value: 500000500000
csfunc via pipeline            00:00:09.8206504    1000000
value: 500000500000
-----------
```

**Reply by AltitudeApps** [reply]
It appears that in this example, there mere presence of a parameter block increases the execution time by a scalar multiple - provided that this parameter accepts pipeline input, and that the function is invoked in the context of a pipeline.

However, if the direct invocation of the function is not done in the immediate context of a pipeline, it starts to balloon in what appears to be an order of magnitude.

**Reply by jimbobmcgee** [reply]
Interesting findings regarding the non-parametised pipeline, but pipelining doesn't always lend itself to a multi-step loop body, which is where I found myself after I'd refactored my original script into functions (e.g. for readability):

````
(1..100) | %{ 
    Do-Thing1 -something -funky $_
    Do-Thing2 -something -else
    Do-Etc
}
````

Also, as an aside, does `$thisMany = 1 .. $script:iter` materialise the numbers into memory, ahead of time, or does it generate them on the fly?  Appreciate that this was an example&mdash;and certainly not relevant to the contrived test&mdash;but I wouldn't want what I was looping through to have to be preloaded upfront.

**Reply by AltitudeApps** [reply]
I haven't checked directly by reading the source, but I believe the  `..` operator, when used in this context and not the "slice" context,  ( slice meaning  `$myArray[ $from .. $to ]` ) yields an array initialized to the sequence of values.  It isn't like a generator function like in javascript; I'm pretty sure the whole value just  _"*poof*!"_ appears in memory when it is evaluated.  (From the point of view of the program text.)   How could it be any other way?  This expression has no idea how you intend to use the value resulting from its evaluation.  Maybe you are going to use or examine each item in this collection one by one, in sequence.  In which case it could produce each item on an "on demand" basis.   Or maybe you are going to choose a random number, and access an item somewhere in the middle of the collection.  In which case, it'd better be there.

So, in our contrived example, we are evaluating the expression `1 .. $script:iter` for each of the `foreach` instances that you have, and for every `$thisMany = 1 .. $script:iter;  $thisMany | & $sbBlahBlah` place where I use it.  So it would make more sense if we put the `$thisMany` declaration and initialization at the top of the main loop, thusly:

```pwsh
foreach ($script:iter in $Iterations) {

    # let's just do this once at the top, or each iteration
    $thisMany = 1 .. $script:iter

    Write-Output (New-Object PSObject -Property @{
    #  blah blah, code continues...  etc. and so forth... 
``` 

But this doesn't answer your question about the "generated on the fly" aspect.  For this you need some sort of iterable gizmo, likely in .NET you'll want something that implements `IEnumerable<X>`, so that you can call `.GetEnumerator()`, which gives you a whatchyamacallit that has implemented whatever logic that is needed to handle the `.MoveNext()` operations required to supply each successive item from whatever set you are providing to the "customer" of this particular object.   In our case, the "customer" will either be a 
`foreach ( $item in $magicalEnumerator ) {}` loop, or a pipeline 
` $magicalEnumerator | & $sbSomeFunctionWhichAcceptsInputViaThePipeline |  & $sbFurtherOperationsOnOutputOfPreviousFunction` . 

I apologize; I'm a bit fuzzy on the details at the moment.  Surely something in the above text is half wrong or misleading, but I hope the the gist of it is clear.  Javascript generators do much the same thing, and it would appear that there has been quite a bit of cross-pollination between the .NET and ECMAScript thinking.  As a consequence I get things mixed up sometimes.

Anyway, to return from my tangent, I'll just end with stating that I have indeed written some code which deals with enormous amounts of data, where "enormous" means "cannot possibly fit in memory", and which uses the Enumerable/Generator like patterns I describe above, to ingest the data chunk by chunk, and passing each chuck down a pipeline.  In this case the pipeline I created is simply used to create a facility which is basically just a composition of number of smaller functions, each being invoked in order, based on the results of the previous function in the sequence.  This code was written entirely in PowerShell, without resorting to C# at any point.

In fact, that's all a pipeline really is:  a really elegant _(if you squint in just the right way)_ and innovative _(no arguing this point)_ way to offer "function composition".  From a conceptual point of view, a pipeline you write in PowerShell can be thought of simply as a larger function (in the computer science sense of the word) which is comprised of a sequence of smaller functions.  _<sub>(Modulo the fact that PowerShell is all _**"YOLO, dude!"**_ about things like referential transparency, and [... insert long list of concepts from functional programming...])</sub>_

So, if your use case involves something like processing data retrieved from an RDBMS, and:
- you are operating on the data one row at a time, and
- you face the prospect that, at times, you'll have so much data that you can't possibly fit it all in memory,

then the Enumerator/Generator pattern would be just what you described as needing.  You obviously already have the code which connects to the database and retrieves the data, what you need now is something that contains that logic, but exposes access to the data by way of an Enumerator.  You can write it in such a way that you grab data in chunks of 1,000 rows, but the enumerator only parcels out one row at a time.  When the rows are gone, it just goes and grabs the next 1,000 rows.


Last thought.   When you write:

```pwsh

(1..100) | %{ 
                                      #  something familiar about this pattern...
    Do-Thing1 -something -funky $_    #  this function is composed of a sequence of operations...
    Do-Thing2 -something -else        #  hmm, "sequence of operations"...  sounds suspiciously 
    Do-Etc                            #  like something I've run across before...  
                                      #  oh, right!   That's what a pipeline is !
}
```
I see this instead:
```pwsh
function Operation1 { process { $fromSource = $_; <# yadda #>; $result  } }
function Operation2 { process { $fromOp1 = $_;    <# yadda #>; $result  } }
function Operation3 { process { $fromOp2 = $_;    <# yadda #>; $result  } }

$manyThings = [SourceOfTruth]::GetAllThings();
$manyThings `
  |               `
    & Operation1  `
  |               `
    & Operation2  `
  |               `
    & Operation3  `
  |               `
    Out-ToWhateverStorage `
  ;
```

Making this observation is a far cry from actually recommending that you switch your implementation to one that uses pipelining.  While the concepts are really simple, once one wraps one's head around them, in actual practice there are many pitfalls and unexpected weird corners to PowerShell.  Such skill and knowledge as I possess, the level of which I leave to the observer to determine, (especially since from my subjective view it seems to swell and contract on a weekly basis), was arrived at only after years of significant pain and frustration.   So, I'm just telling you two things:
1. that it's _possible_ to implement more than you might expect in terms of "pipeline-able" functions, and
2. in my experience, PowerShell really _wants_ pipelines to be used; it is one of the fundamental concepts at the core of the language.  

Whether or not you choose to embark on this adventure is not something I should offer any input on.

Just consider how much faster the  
`function foo { process {  $c = $_.a + $_.b;  $_.c = $c; $_  } };  $manyABitems | & foo` 
approach executed, compared to the supposed equivalents.  PowerShell is a thoroughbred, but it wants to be ridden bareback.  You only need a bridle, helmet and goggles.  Hop on and hold on for dear life!

**Reply by AltitudeApps** [reply]
P.S.  When I use the word "function", I usually almost always mean "scriptblock".  This is because the "scriptblock" is PowerShell's idea of the "anonymous function".   The anonymous function is a concept central to many languages.   So much so that in my head when I hear "function", I think "anonymous function", which maps directly to "scriptblock" in the context of PowerShell.

Also, calls to `$scriptblock.GetNewClosure()` will fail in hilarious and entertaining ways if you use this method on a scriptblock that uses named functions anywhere in its body, instead of only invoking scriptblocks directly via one of the invocation operators.  ( `.` and `&` )     If I spend any significant time away from PowerShell, I get burned by that when I return.   Every.  Single.   Time.

So any code that I write that isn't "client facing" is usually written exclusively in terms of scriptblocks.

**Comment by iSazonov** [comment]
Parameter binding area is very complex and sensitive. That's the price of versatility.

The current implementation of parameter bindings is well optimized. It is unlikely that we can do anything better there directly.
Nevertheless, I see 3 directions for work:
1. recognition of well-known types
2. caching of binding results (this may already be being done)
3. benefits from compiling based on meta data (e.g., OutputType attribute) (this is similar to C# source generators - take meta data, convert to code, compile, fast use)

Unfortunately, we do not have community experts who would implement improvements in this area.
